src/cdk/stepper/stepper.ts

```diff
@@ -0,0 +1,153 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+    Component, ContentChildren, EventEmitter, Input, Output, QueryList, OnInit,
+    AfterViewChecked, AfterViewInit, Directive, ElementRef, ViewChild, ViewChildren
+} from '@angular/core';
+import {CdkStep} from './step';
+import {Observable} from 'rxjs/Observable';
+import {map} from 'rxjs/operator/map';
+import {LEFT_ARROW, RIGHT_ARROW, ENTER, TAB} from '../keyboard/keycodes';
+
+/** Used to generate unique ID for each stepper component. */
+let nextId = 0;
+
+/** Change event emitted on focus or selection changes. */
+export class CdkStepEvent {
+    index: number;
+    step: CdkStep;
+}
+
+@Directive({
+    selector: '[cdkStepper]',
+    host: {
+        '(keydown)': '_onKeydown($event)'
+    },
+})
+export class CdkStepper {
+    @ContentChildren(CdkStep) _steps: QueryList<CdkStep>;
+
+    @ViewChildren('stepHeader') _stepHeader: QueryList<ElementRef>;
+
+    /** The index of the currently selected step. */
+    @Input()
+    set selectedIndex(value: number) {
+        this._selectedIndex = value;
+    }
+    get selectedIndex(): number { return this._selectedIndex; }
+    private _selectedIndex: number;
+
+    /** Optional input to support both linear and non-linear stepper component. */
+    @Input() linear: boolean = true;
+
+    /** Output to enable support for two-way binding on `[(selectedIndex)]` */
+    @Output() get selectedIndexChange(): Observable<number> {
+        return map.call(this.stepEvent, event => event.index);
+    }
+
+    // @Output() get focusIndexChange(): Observable<number> {
+    //     return map.call(this.focusChange, event => event.index);
+    // }
+
+    /** Event emitted when the selected step has changed. */
+    @Output() stepEvent = new EventEmitter<CdkStepEvent>();
+
+    /** Event emitted when the focused step has changed. */
+    @Output() focusChange = new EventEmitter<CdkStepEvent>();
+
+    /** The step that is currently selected. */
+    get selectedStep(): CdkStep {
+        return this._steps.toArray()[this._selectedIndex];
+    }
+    private _selectedStep: CdkStep;
+
+    /** The index of the step that the focus is currently on. */
+    get focusIndex(): number {return this._focusIndex; }
+    private _focusIndex: number = 0;
+
+    private _groupId: number;
+
+    constructor() {
+        this._groupId = nextId++;
+    }
+
+    /** Selects and focuses the provided step. */
+    selectStep(step: CdkStep): void {
```

Comment:
I'm saying that as an API I prefer: `step.select()` to `stepper.select(step)`, but this would require the step to know about its parent stepper, so maybe not worth it. @jelbourn do you have any opinion?